azure-communication-ui/chat/src/main/java/com/azure/android/communication/ui/chat/repository/MessageRepositoryListStorage.kt

```diff
@@ -0,0 +1,77 @@
+package com.azure.android.communication.ui.chat.repository
+
+import com.azure.android.communication.ui.chat.models.EMPTY_MESSAGE_INFO_MODEL
+import com.azure.android.communication.ui.chat.models.MessageInfoModel
+import java.util.Collections
+
+internal class MessageRepositoryListWriter : MessageRepositoryWriter {
+    // Simple List for now
+    val messages: MutableList<MessageInfoModel> = Collections.synchronizedList(mutableListOf<MessageInfoModel>())
+
+    // Middleware Interface
+    override fun addLocalMessage(messageInfoModel: MessageInfoModel) {
+        messages.add(messageInfoModel)
+        reorder()
+    }
+
+    override fun addPage(page: List<MessageInfoModel>) {
+        messages.addAll(0, page)
+        reorder()
+    }
+
+    override fun addServerMessage(message: MessageInfoModel) {
+        messages.add(message)
+        reorder()
+    }
+
+    override fun removeMessage(message: MessageInfoModel) {
+        messages.retainAll { it.id != message.id }
+    }
+
+    override fun editMessage(message: MessageInfoModel) {
+        val idx = messages.indexOfFirst {
+            it.id == message.id
+        }
+
+        if (idx != -1) {
+            // TODO: Merge with old message, keep metadata such as type. Update the old message with new message contents
+            mergeWithPreviousMessage(idx, message)
+        }
+        reorder()
+    }
+
+    override fun getLastMessage(): MessageInfoModel? = messages?.last()
+
+    fun reorder() {
+        // TODO: Will need to update with repository stable algorithm implementation
+        messages.sortBy {
+            it.createdOn?.nano
+        }
+    }
+
+    fun mergeWithPreviousMessage(idx: Int, message: MessageInfoModel) {
+        var newMessage = MessageInfoModel(
+            id = messages[idx].id,
+            internalId = messages[idx].internalId,
+            content = message.content,
+            messageType = messages[idx].messageType,
+            version = messages[idx].version,
+            senderDisplayName = messages[idx].senderDisplayName,
+            createdOn = messages[idx].createdOn,
+            editedOn = messages[idx].editedOn,
+            deletedOn = messages[idx].deletedOn,
+            senderCommunicationIdentifier = messages[idx].senderCommunicationIdentifier
+        )
+        messages[idx] = newMessage
+    }
+}
+
+internal class MessageRepositoryListReader(private val writer: MessageRepositoryListWriter) : MessageRepositoryReader() {
```

Comment:
To access repository we need to pass writerImplementation? 
Can we make it simpler? 
I think previous architecture was much simpler and easy to adapt for any new implementation of the repository.

Reply:
we can setup defaults in the constructor if you like.

Since reader takes writer, I didn't bother, but can be done with a factory constructor

Reply:
that way we can have multiple factories.

e.g.  MessageRepository.ListImplementation()
         MessageRepository.TreeImplementation()
         
         etc.

Reply:
I'll update

Reply:
Thanks Adam for the changes